source/gmsh/utilities.cc

```diff
@@ -144,15 +155,201 @@ namespace Gmsh
   }
 #  endif
 
+#  ifdef DEAL_II_GMSH_WITH_API
+  template <int dim, int spacedim>
+  void
+  read_partitioned_msh(
+    parallel::fullydistributed::Triangulation<dim, spacedim> &tria,
+    const MPI_Comm                                            mpi_comm,
+    const std::string                                        &file_prefix,
+    const std::string                                        &file_suffix)
+  {
+    const unsigned int nprocs = Utilities::MPI::n_mpi_processes(mpi_comm);
+    const unsigned int rank   = Utilities::MPI::this_mpi_process(mpi_comm);
+
+    std::string fname =
+      file_prefix + "_" + std::to_string(rank + 1) + "." + file_suffix;
+
+    if (nprocs == 1)
+      {
+        fname = file_prefix + "." + file_suffix;
+        std::ifstream f(fname);
+        AssertThrow(f.good(), ExcMessage("Missing mesh file: " + fname));
+      }
+    else
+      {
+        for (unsigned int i = 1; i <= nprocs; ++i)
+          {
+            const std::string check_fname =
+              file_prefix + "_" + std::to_string(i) + "." + file_suffix;
+            std::ifstream f(check_fname);
+            AssertThrow(f.good(),
+                        ExcMessage("Missing mesh file: " + check_fname));
+          }
+
+        const std::string extra_fname =
+          file_prefix + "_" + std::to_string(nprocs + 1) + "." + file_suffix;
+        std::ifstream f(extra_fname);
+        AssertThrow(!f.good(),
+                    ExcMessage("Unexpected extra mesh file(s): " +
+                               extra_fname));
+      }
+
+    const std::map<int, std::uint8_t> gmsh_to_dealii_type = {
+      {15, 0}, {1, 1}, {2, 2}, {3, 3}, {4, 4}, {7, 5}, {6, 6}, {5, 7}};
+
+    const std::array<std::vector<unsigned int>, 8> gmsh_to_dealii = {
+      {{0},
+       {0, 1},
+       {0, 1, 2},
+       {0, 1, 3, 2},
+       {0, 1, 2, 3},
+       {0, 1, 3, 2, 4},
+       {0, 1, 2, 3, 4, 5},
+       {0, 1, 3, 2, 4, 5, 7, 6}}};
+
+    gmsh::initialize();
+    gmsh::option::setNumber("General.Verbosity", 0);
+    gmsh::open(fname);
```

Comment:
What happens if `gmsh` has *already* been initialized? What if someone has specifically set the verbosity to something other than 0? I wonder whether we need to run these two lines somewhere else, in the initialization of deal.II itself.

Reply:
no, this needs to be done everytime you want to work on a mesh. Pretty much like mpi init/finalize, so every function that reads a gmsh file needs to do this separately.

Reply:
I see. Then we should probably guard all of these places with a mutex to avoid getting into threadsafety issues. That's of course a downstream thing, not for this patch.